File: agent/src/cinepais_agent/events.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, model_validator

Format = Literal["IMAX", "Onyx", "2D", "Doblada", "Subtitulada", "Premium"]


class TokenEvent(BaseModel):
    type: Literal["token"] = "token"
    content: str


class ToolCallEvent(BaseModel):
    type: Literal["tool_call"] = "tool_call"
    tool: str
    input: dict  # type: ignore[type-arg]


class Alternative(BaseModel):
    """An alternative showtime option to present alongside the main recommendation."""

    showtimeId: str
    filmId: str
    siteName: str
    businessDate: str  # YYYY-MM-DD
    time: str  # HH:MM
    formats: list[Format]
    priceFrom: int
    qualityTier: Literal["low", "optimal", "high"] | None
    reason: str  # Spanish reason, e.g. "mejor calidad de silla"


class RecommendationEvent(BaseModel):
    type: Literal["recommendation"] = "recommendation"
    outcome: Literal["recommended", "degraded", "no_availability"]
    showtimeId: str | None
    filmId: str | None
    seatIds: list[str]
    requestedN: int
    siteName: str | None
    city: str | None
    businessDate: str | None
    time: str | None
    formats: list[Format]
    priceFrom: int | None
    qualityTier: Literal["low", "optimal", "high"] | None
    reasoning: str
    alternatives: list[Alternative] = []

    @model_validator(mode="after")
    def validate_outcome_contract(self) -> RecommendationEvent:
        """Enforce the outcome branch contract."""
        if self.outcome == "recommended":
            if self.showtimeId is None:
                raise ValueError("recommended outcome requires showtimeId")
            if len(self.seatIds) != self.requestedN:
                raise ValueError(
                    f"recommended outcome requires len(seatIds)==requestedN, "
                    f"got {len(self.seatIds)} != {self.requestedN}"
                )
        elif self.outcome == "degraded":
            if self.showtimeId is None:
                raise ValueError("degraded outcome requires showtimeId")
            if not (1 <= len(self.seatIds) < self.requestedN):
                raise ValueError(
                    f"degraded outcome requires 1<=len(seatIds)<requestedN, "
                    f"got {len(self.seatIds)}, requestedN={self.requestedN}"
                )
        elif self.outcome == "no_availability":
            if self.showtimeId is not None:
                raise ValueError("no_availability outcome requires showtimeId to be None")
            if self.seatIds:
                raise ValueError("no_availability outcome requires seatIds to be []")
        return self


class DoneEvent(BaseModel):
    type: Literal["done"] = "done"
    sessionQueriesUsed: int
    sessionQueryCap: int


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str  # Spanish message

File: web/src/lib/agent/events.ts
import { z } from "zod";

/**
 * Zod mirror of the agent's SSE event contract.
 *
 * Source of truth: `agent/src/cinepais_agent/events.py` — these schemas mirror
 * those Pydantic models field-for-field. Nullability is copied literally:
 * `Alternative.priceFrom` is `int` (never null) while
 * `RecommendationEvent.priceFrom` is `int | None`, because a
 * `no_availability` outcome carries no price.
 *
 * `formats` is typed as `string[]` rather than the `Format` enum on purpose:
 * the agent already validates it against `Literal[...]` before it reaches the
 * wire, and a permissive array here means a future format never turns a whole
 * recommendation into a dropped frame.
 */

const QualityTierSchema = z.enum(["low", "optimal", "high"]);

export const TokenEventSchema = z.object({
  type: z.literal("token"),
  content: z.string(),
});

export const ToolCallEventSchema = z.object({
  type: z.literal("tool_call"),
  tool: z.string(),
  input: z.record(z.string(), z.unknown()),
});

/** An alternative showtime presented alongside the main recommendation. */
export const AlternativeSchema = z.object({
  showtimeId: z.string(),
  filmId: z.string(),
  siteName: z.string(),
  businessDate: z.string(),
  time: z.string(),
  formats: z.array(z.string()),
  // events.py:30 — `int`, NOT `int | None`.
  priceFrom: z.number(),
  // events.py:31 — null marks a soldout alternative.
  qualityTier: QualityTierSchema.nullable(),
  reason: z.string(),
});

export const RecommendationEventSchema = z.object({
  type: z.literal("recommendation"),
  outcome: z.enum(["recommended", "degraded", "no_availability"]),
  showtimeId: z.string().nullable(),
  filmId: z.string().nullable(),
  seatIds: z.array(z.string()),
  requestedN: z.number(),
  siteName: z.string().nullable(),
  city: z.string().nullable(),
  businessDate: z.string().nullable(),
  time: z.string().nullable(),
  formats: z.array(z.string()),
  // events.py:47 — `int | None`. A `no_availability` outcome sends null here.
  priceFrom: z.number().nullable(),
  qualityTier: QualityTierSchema.nullable(),
  reasoning: z.string(),
  alternatives: z.array(AlternativeSchema).default([]),
});

export const DoneEventSchema = z.object({
  type: z.literal("done"),
  sessionQueriesUsed: z.number(),
  sessionQueryCap: z.number(),
});

export const ErrorEventSchema = z.object({
  type: z.literal("error"),
  code: z.string(),
  message: z.string(),
});

export const AgentEventSchema = z.discriminatedUnion("type", [
  TokenEventSchema,
  ToolCallEventSchema,
  RecommendationEventSchema,
  DoneEventSchema,
  ErrorEventSchema,
]);

// Type exports
export type QualityTier = z.infer<typeof QualityTierSchema>;
export type TokenEvent = z.infer<typeof TokenEventSchema>;
export type ToolCallEvent = z.infer<typeof ToolCallEventSchema>;
export type Alternative = z.infer<typeof AlternativeSchema>;
export type RecommendationEvent = z.infer<typeof RecommendationEventSchema>;
export type DoneEvent = z.infer<typeof DoneEventSchema>;
export type ErrorEvent = z.infer<typeof ErrorEventSchema>;
export type AgentEvent = z.infer<typeof AgentEventSchema>;

/** The SSE event names the agent emits. */
const KNOWN_EVENT_NAMES = new Set<string>([
  "token",
  "tool_call",
  "recommendation",
  "done",
  "error",
]);

/**
 * Parse one SSE frame into a typed agent event.
 *
 * Returns `null` — never throws — when the payload is malformed JSON, fails
 * the schema, or carries a `type` that contradicts a known SSE event name.
 * One bad frame must never kill the stream.
 *
 * `eventName` is advisory: an unrecognized name (including the WHATWG default
 * `"message"`) defers entirely to the payload's own `type` discriminant.
 */
export function parseAgentEvent(
  eventName: string,
  rawData: string,
): AgentEvent | null {
  let json: unknown;
  try {
    json = JSON.parse(rawData);
  } catch {
    return null;
  }

  const result = AgentEventSchema.safeParse(json);
  if (!result.success) return null;

  if (KNOWN_EVENT_NAMES.has(eventName) && result.data.type !== eventName) {
    return null;
  }

  return result.data;
}
